drivers/wireless/cyw43/bus.zig

```diff
@@ -0,0 +1,370 @@
+            const words_to_send = (len + 3) / 4 + 1;
+            _ = self.spi.spi_read_blocking(@bitCast(cmd), buf[0..words_to_send]);
+
+            const u32_data_slice = buf[1..];
+            var u8_buf_view = std.mem.sliceAsBytes(u32_data_slice);
+
+            @memcpy(remaining_data[0..len], u8_buf_view[0..len]);
+
+            current_addr += @as(u32, @intCast(len));
+            remaining_data = remaining_data[len..];
+        }
+    }
+
+    pub fn bp_read8(self: *Self, addr: u32) u8 {
+        return @truncate(self.backplane_readn(addr, 1));
+    }
+
+    pub fn bp_read16(self: *Self, addr: u32) u16 {
+        return @truncate(self.backplane_readn(addr, 2));
+    }
+
+    pub fn bp_read32(self: *Self, addr: u32) u32 {
+        return @truncate(self.backplane_readn(addr, 4));
+    }
+
+    pub fn backplane_readn(self: *Self, addr: u32, len: u11) u32 {
+        log.debug("backplane_readn addr = 0x{X} len = {}", .{ addr, len });
+
+        self.backplane_set_window(addr);
+
+        var bus_addr = addr & consts.BACKPLANE_ADDRESS_MASK;
+        if (len == 4) {
+            bus_addr |= consts.BACKPLANE_ADDRESS_32BIT_FLAG;
+        }
+
+        const val = self.readn(.backplane, @truncate(bus_addr), len);
+
+        log.debug("backplane_readn addr = 0x{X} len = {} val = 0x{X}", .{ addr, len, val });
+
+        return val;
+    }
+
+    pub fn bp_write(self: *Self, addr: u32, data: []const u8) void {
+        log.debug("bp_write addr = 0x{X} len = {}", .{ addr, data.len });
+
+        // It seems the HW force-aligns the addr
+        // to 2 if data.len() >= 2
+        // to 4 if data.len() >= 4
+        // To simplify, enforce 4-align for now.
+        std.debug.assert(addr % 4 == 0);
+
+        var buf: [consts.BACKPLANE_MAX_TRANSFER_SIZE / 4 + 1]u32 = undefined;
+
+        var current_addr = addr;
+        var remaining_data = data;
+
+        while (remaining_data.len > 0) {
+            const window_offs = current_addr & consts.BACKPLANE_ADDRESS_MASK;
+            const window_remaining = consts.BACKPLANE_WINDOW_SIZE - @as(usize, @intCast(window_offs));
+
+            const len: usize = @min(remaining_data.len, consts.BACKPLANE_MAX_TRANSFER_SIZE, window_remaining);
+
+            const u32_data_slice = buf[1..];
+            var u8_buf_view = std.mem.sliceAsBytes(u32_data_slice);
+
+            @memcpy(u8_buf_view[0..len], remaining_data[0..len]);
+
+            self.backplane_set_window(current_addr);
+
+            const cmd = Cyw43Cmd{ .cmd = .write, .incr = .incremental, .func = .backplane, .addr = @truncate(window_offs), .len = @truncate(len) };
+            buf[0] = @bitCast(cmd);
+
+            const words_to_send = (len + 3) / 4 + 1;
+            _ = self.spi.spi_write_blocking(buf[0..words_to_send]);
+
+            current_addr += @as(u32, @intCast(len));
+            remaining_data = remaining_data[len..];
+        }
+    }
+
+    pub fn bp_write8(self: *Self, addr: u32, value: u8) void {
```

Comment:
inline like other n variants (or inline none and count on compiler. you could make the len field comptime probably)

Reply:
I will go with inline for now so it will be easy to keep follow it with original reference code from Rust.